core/trino-main/src/main/java/io/trino/operator/window/pattern/SetEvaluator.java

```diff
@@ -0,0 +1,175 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.window.pattern;
+
+import io.trino.operator.window.matcher.ArrayView;
+import io.trino.operator.window.matcher.IntList;
+import io.trino.sql.planner.rowpattern.AggregatedSetDescriptor;
+import io.trino.sql.planner.rowpattern.ir.IrLabel;
+
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class returns a set of positions to aggregate over for an aggregate function in row pattern matching context.
+ * Aggregations in row pattern matching context have RUNNING or FINAL semantics, and they apply only to rows
+ * matched with certain pattern variables. For example, for a match "A B A A B", the aggregation `sum(B.x)` only
+ * applies to the second and the last position.
+ * <p>
+ * This evaluator is stateful. It requires a reset for every new match.
+ * The method `resolveNewPositions()` returns a portion of positions corresponding to the new portion of the match
+ * since the last call. It is thus assumed that a sequence of calls since the instance creation or `reset()`
+ * applies to the same match (i.e. the `matchedLabels` passed as an argument to one call is prefix of
+ * `matchedLabels` passed in the next call).
+ * <p>
+ * Also, a full list of positions for a current match is kept, and it can be obtained via the `getAllPositions()`
+ * method. This is for the purpose comparing pattern matching threads in ThreadEquivalence.
+ */
+public class SetEvaluator
+{
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private final Set<Integer> labels;
+    private final boolean running;
+
+    // length of the aggregated prefix of the match
+    private int aggregated;
+
+    // length of the prefix of the match where all applicable positions were identified,
+    // and stored in `allPositions`. During the pattern matching phase it might exceed
+    // the `aggregated` prefix due to `getAllPositions()` calls.
+    private int evaluated;
+
+    // all identified applicable positions for the current match until the `evaluated` position, starting from 0
+    // this list is updated:
+    // - by `resolveNewPositions()`
+    // - by `getAllPositions()` (only during the pattern matching phase)
+    private final IntList allPositions;
```

Comment:
It's unclear what aggregated vs evaluated vs allPositions represent. I guess part of my confusion is due to not understanding how this class fits in the overall picture. Can you elaborate a bit?

Reply:
The `SetEvaluator` is part of the state of `MatchAggregation`. 
Its main purpose is to provide new applicable positions(*) to be aggregated by the MatchAggregation, since the previous call.
The additional purpose is to keep a collection of all applicable positions to be used by ThreadEquivalence.

Accordingly to the two purposes, two pointers are used to mark how much input was processed:
- `aggregated` indicates how much input was processed by the MatchAggregation,
- `evaluated` indicates how much input was processed to find applicable positions.

`evaluated` is never behind `aggregated`, because if we run the aggregation on some portion of input, we must have resolved the applicable positions first.
`evaluated` might be ahead of `aggregated`, in the case when we resolved the positions, but didn't run the actual aggregation. That happens via the `getAllPositions()` call, when ThreadEquivalence asks for the set of positions and compares them, but doesn't run the actual aggregation.

- `allPositions` are all the applicable positions up until the `evaluated` pointer

(*)  applicable positions are the positions with certain labels, e.g. for `avg(A.x)` these will be the positions labeled with `A` in the current match, wrt the running or final semantics.